File: app/lib/data/widget/widget_snapshot.dart
import '../models/card.dart';
import '../models/home_response.dart';
import '../models/json.dart';

/// The compact payload the Android home-screen widget renders (docs/06 §Home-screen widget).
///
/// The widget draws with `RemoteViews`, outside the Flutter engine, so everything it shows has
/// to survive as plain JSON in shared storage. Two producers write this exact shape and both
/// must stay in step:
///   * Dart — [WidgetSnapshot.fromHome], on every `/home` payload the app receives;
///   * Kotlin — `MausamWidgetWorker` / `SnapshotJson.kt`, on the hourly background refresh.
///
/// Card *content* is already localized by the backend (docs/04 `lang`), so nothing here is
/// translated on the device; [lang] only records which language the strings are in, so the
/// widget can tell a stale Hindi snapshot from a fresh English one after a language switch.
class WidgetSnapshot {
  const WidgetSnapshot({
    this.locationName = '',
    this.updatedAt,
    this.lang = 'en',
    this.units = 'metric',
    this.hero,
    this.pinned,
  });

  /// docs/04 §Objects `HomeResponse.location.name`.
  final String locationName;

  /// When the data was produced, not when it was written: the payload's newest `freshness`
  /// timestamp, falling back to `generated_at`. The widget's age label and its ">6 h → open
  /// Mausam to refresh" rule are both computed from this, so a cache replay stays honest.
  final String? updatedAt;

  final String lang;

  /// `metric` | `imperial` — the widget formats the temperature itself.
  final String units;

  final WidgetHero? hero;

  /// The top pinned card, or the top ranked card when nothing is pinned.
  final WidgetPinned? pinned;

  /// Nothing worth drawing: the widget falls back to "Open Mausam to refresh".
  bool get isEmpty => hero == null && pinned == null;

  /// Empty payloads still carry the language so the widget's own chrome keeps its locale.
  static const WidgetSnapshot empty = WidgetSnapshot();

  /// Builds the snapshot from a `/home` (or `/home?lite=1`) response.
  ///
  /// Selection rules, mirrored in Kotlin:
  ///   * hero  — `hero`, which docs/02 card 1 guarantees is `current_conditions`;
  ///   * pinned — `pinned.first` when the ranker pinned anything (a warning always is),
  ///     otherwise the first ranked card, which is the highest scoring one.
  factory WidgetSnapshot.fromHome(
    HomeResponse home, {
    String? lang,
    String units = 'metric',
  }) {
    final top = home.pinned.isNotEmpty
        ? home.pinned.first
        : (home.cards.isNotEmpty ? home.cards.first : null);
    return WidgetSnapshot(
      locationName: home.location.name,
      updatedAt: home.newestFreshness ?? home.generatedAt,
      lang: lang ?? home.context.lang,
      units: units,
      hero: home.hero == null ? null : WidgetHero.fromCard(home.hero!),
      pinned: top == null ? null : WidgetPinned.fromCard(top),
    );
  }

  factory WidgetSnapshot.fromJson(Map<String, dynamic> json) => WidgetSnapshot(
        locationName: asString(json['location']),
        updatedAt: asStringOrNull(json['updated_at']),
        lang: asString(json['lang'], fallback: 'en'),
        units: asString(json['units'], fallback: 'metric'),
        hero: json['hero'] == null ? null : WidgetHero.fromJson(asMap(json['hero'])),
        pinned: json['pinned'] == null ? null : WidgetPinned.fromJson(asMap(json['pinned'])),
      );

  /// `v` is the schema version: the Kotlin side ignores a snapshot it does not understand
  /// rather than drawing half of it.
  Map<String, dynamic> toJson() => <String, dynamic>{
        'v': schemaVersion,
        'location': locationName,
        'updated_at': updatedAt,
        'lang': lang,
        'units': units,
        'hero': hero?.toJson(),
        'pinned': pinned?.toJson(),
      };

  static const int schemaVersion = 1;
}

/// docs/02 card 1 `current_conditions` — the four things that fit on one widget row.
class WidgetHero {
  const WidgetHero({
    this.tempC,
    this.condition = '',
    this.icon,
    this.feelsLikeC,
    this.tmaxC,
    this.tminC,
  });

  final double? tempC;

  /// Already localized by the backend (`data.condition_text`).
  final String condition;

  /// docs/06 §Layout `core/icons.dart` icon name, e.g. `thunderstorm`.
  final String? icon;
  final double? feelsLikeC;
  final double? tmaxC;
  final double? tminC;

  factory WidgetHero.fromCard(HomeCard card) {
    final d = card.data;
    return WidgetHero(
      tempC: asDouble(d['temp_c']),
      condition: asString(d['condition_text'], fallback: card.subtitle),
      icon: asStringOrNull(d['icon']),
      feelsLikeC: asDouble(d['feels_like_c']),
      tmaxC: asDouble(d['tmax_c']),
      tminC: asDouble(d['tmin_c']),
    );
  }

  factory WidgetHero.fromJson(Map<String, dynamic> json) => WidgetHero(
        tempC: asDouble(json['temp_c']),
        condition: asString(json['condition']),
        icon: asStringOrNull(json['icon']),
        feelsLikeC: asDouble(json['feels_like_c']),
        tmaxC: asDouble(json['tmax_c']),
        tminC: asDouble(json['tmin_c']),
      );

  Map<String, dynamic> toJson() => <String, dynamic>{
        'temp_c': tempC,
        'condition': condition,
        'icon': icon,
        'feels_like_c': feelsLikeC,
        'tmax_c': tmaxC,
        'tmin_c': tminC,
      };
}

/// One line of "what matters right now": the pinned card's title, its insight headline and the
/// colour the widget paints its accent bar with.
class WidgetPinned {
  const WidgetPinned({
    required this.type,
    this.title = '',
    this.insight = '',
    this.severity = 'info',
    this.colorHex = WidgetPinned.infoHex,
    this.estimated = false,
  });

  final String type;
  final String title;

  /// `insight.headline`, falling back to the subtitle — one line, already localized.
  final String insight;

  /// info|advisory|watch|warning|severe (docs/02 §Card anatomy).
  final String severity;

  /// `#RRGGBB`. For a `warnings` card this is the IMD warning colour that came with the
  /// payload, so the widget shows the same yellow/orange/red as the app.
  final String colorHex;

  /// CLAUDE.md §6 — modelled data is labelled as such, in the widget too.
  final bool estimated;

  /// docs/06 §Layout `core/theme.dart` severity colours, as hex so Kotlin can parse them.
  static const String yellowHex = '#F5C518';
  static const String orangeHex = '#F28C28';
  static const String redHex = '#D32F2F';

  /// The M3 seed (#1565C0) stands in for `colorScheme.primary`, which a RemoteViews layout
  /// cannot resolve.
  static const String infoHex = '#1565C0';

  factory WidgetPinned.fromCard(HomeCard card) => WidgetPinned(
        type: card.type,
        title: card.title,
        insight: card.insight?.headline.isNotEmpty == true
            ? card.insight!.headline
            : card.subtitle,
        severity: card.severity,
        colorHex: colorFor(card),
        estimated: card.isEstimated,
      );

  /// A `warnings` card is coloured by the warning it carries (docs/04 `Warning.color_hex`);
  /// everything else by the card severity band (docs/06 §theme `cardSeverityColor`).
  static String colorFor(HomeCard card) {
    if (card.type == 'warnings') {
      final warnings = card.data['warnings'];
      if (warnings is List && warnings.isNotEmpty && warnings.first is Map) {
        final first = asMap(warnings.first);
        final hex = asStringOrNull(first['color_hex']);
        if (hex != null && hex.isNotEmpty) return hex;
        switch (asStringOrNull(first['severity'])) {
          case 'red':
            return redHex;
          case 'orange':
            return orangeHex;
          case 'yellow':
            return yellowHex;
        }
      }
    }
    switch (card.severity) {
      case 'severe':
        return redHex;
      case 'warning':
        return orangeHex;
      case 'watch':
        return yellowHex;
      default:
        return infoHex;
    }
  }

  factory WidgetPinned.fromJson(Map<String, dynamic> json) => WidgetPinned(
        type: asString(json['type']),
        title: asString(json['title']),
        insight: asString(json['insight']),
        severity: asString(json['severity'], fallback: 'info'),
        colorHex: asString(json['color_hex'], fallback: infoHex),
        estimated: asBool(json['estimated']),
      );

  Map<String, dynamic> toJson() => <String, dynamic>{
        'type': type,
        'title': title,
        'insight': insight,
        'severity': severity,
        'color_hex': colorHex,
        'estimated': estimated,
      };
}

File: app/android/app/src/main/kotlin/com/teammausam/mausam_app/WidgetSnapshot.kt
package com.teammausam.mausam_app

import org.json.JSONObject

/**
 * The compact payload the widget draws, and the two ways it gets built.
 *
 * The schema is defined once in `app/lib/data/widget/widget_snapshot.dart`; this file is its
 * Kotlin twin, because the hourly background refresh runs with no Flutter engine attached and
 * has to produce byte-compatible JSON from a raw `/home?lite=1` response.
 *
 * Card content is never translated here — it arrives already localized from the backend
 * (docs/04 `lang`).
 */
data class WidgetSnapshot(
    val locationName: String,
    val updatedAt: String?,
    val lang: String,
    val units: String,
    val hero: Hero?,
    val pinned: Pinned?,
) {
    data class Hero(
        val tempC: Double?,
        val condition: String,
        val feelsLikeC: Double?,
        val tmaxC: Double?,
        val tminC: Double?,
    )

    data class Pinned(
        val type: String,
        val title: String,
        val insight: String,
        val severity: String,
        val colorHex: String,
        val estimated: Boolean,
    )

    val isEmpty: Boolean get() = hero == null && pinned == null

    companion object {
        const val SCHEMA_VERSION = 1

        private const val INFO_HEX = "#1565C0"
        private const val YELLOW_HEX = "#F5C518"
        private const val ORANGE_HEX = "#F28C28"
        private const val RED_HEX = "#D32F2F"

        val EMPTY = WidgetSnapshot("", null, "en", "metric", null, null)

        /** Reads what the app (or a previous worker run) stored. Never throws. */
        fun parse(raw: String?): WidgetSnapshot {
            if (raw.isNullOrBlank()) return EMPTY
            return try {
                val json = JSONObject(raw)
                if (json.optInt("v", SCHEMA_VERSION) > SCHEMA_VERSION) return EMPTY
                val heroJson = json.optJSONObject("hero")
                val pinnedJson = json.optJSONObject("pinned")
                WidgetSnapshot(
                    locationName = json.optString("location", ""),
                    updatedAt = json.optStringOrNull("updated_at"),
                    lang = json.optString("lang", "en"),
                    units = json.optString("units", "metric"),
                    hero = heroJson?.let {
                        Hero(
                            tempC = it.optDoubleOrNull("temp_c"),
                            condition = it.optString("condition", ""),
                            feelsLikeC = it.optDoubleOrNull("feels_like_c"),
                            tmaxC = it.optDoubleOrNull("tmax_c"),
                            tminC = it.optDoubleOrNull("tmin_c"),
                        )
                    },
                    pinned = pinnedJson?.let {
                        Pinned(
                            type = it.optString("type", ""),
                            title = it.optString("title", ""),
                            insight = it.optString("insight", ""),
                            severity = it.optString("severity", "info"),
                            colorHex = it.optString("color_hex", INFO_HEX),
                            estimated = it.optBoolean("estimated", false),
                        )
                    },
                )
            } catch (e: Exception) {
                EMPTY
            }
        }

        /**
         * Builds a snapshot from a `/home` (or `/home?lite=1`) response, with the same selection
         * rules as `WidgetSnapshot.fromHome` in Dart: the `hero` card, plus `pinned.first` when
         * anything is pinned and otherwise the top ranked card.
         */
        fun fromHome(home: JSONObject, lang: String, units: String): WidgetSnapshot {
            val pinnedArray = home.optJSONArray("pinned")
            val cards = home.optJSONArray("cards")
            val top = when {
                pinnedArray != null && pinnedArray.length() > 0 -> pinnedArray.optJSONObject(0)
                cards != null && cards.length() > 0 -> cards.optJSONObject(0)
                else -> null
            }
            val heroCard = home.optJSONObject("hero")
            val heroData = heroCard?.optJSONObject("data")
            return WidgetSnapshot(
                locationName = home.optJSONObject("location")?.optString("name", "") ?: "",
                updatedAt = newestFreshness(home),
                lang = lang,
                units = units,
                hero = heroCard?.let {
                    Hero(
                        tempC = heroData?.optDoubleOrNull("temp_c"),
                        condition = heroData?.optString("condition_text", "")?.ifEmpty {
                            it.optString("subtitle", "")
                        } ?: it.optString("subtitle", ""),
                        feelsLikeC = heroData?.optDoubleOrNull("feels_like_c"),
                        tmaxC = heroData?.optDoubleOrNull("tmax_c"),
                        tminC = heroData?.optDoubleOrNull("tmin_c"),
                    )
                },
                pinned = top?.let {
                    val headline = it.optJSONObject("insight")?.optString("headline", "").orEmpty()
                    Pinned(
                        type = it.optString("type", ""),
                        title = it.optString("title", ""),
                        insight = headline.ifEmpty { it.optString("subtitle", "") },
                        severity = it.optString("severity", "info"),
                        colorHex = colorFor(it),
                        estimated = it.optBoolean("estimated", false) ||
                            it.optString("source", "") == "estimated",
                    )
                },
            )
        }

        /** The newest of `freshness.*`, falling back to `generated_at` (mirrors the Dart model). */
        private fun newestFreshness(home: JSONObject): String? {
            val freshness = home.optJSONObject("freshness")
            var best: String? = null
            var bestAt: Long = Long.MIN_VALUE
            if (freshness != null) {
                for (key in freshness.keys()) {
                    val value = freshness.optStringOrNull(key) ?: continue
                    val at = WidgetTime.parseIso(value) ?: continue
                    if (at > bestAt) {
                        bestAt = at
                        best = value
                    }
                }
            }
            return best ?: home.optStringOrNull("generated_at")
        }

        /**
         * A `warnings` card takes the IMD colour of the warning it carries (docs/04
         * `Warning.color_hex`); anything else the severity band from docs/02 §Card anatomy.
         */
        fun colorFor(card: JSONObject): String {
            if (card.optString("type", "") == "warnings") {
                val first = card.optJSONObject("data")?.optJSONArray("warnings")?.optJSONObject(0)
                if (first != null) {
                    val hex = first.optStringOrNull("color_hex")
                    if (hex != null) return hex
                    when (first.optString("severity", "")) {
                        "red" -> return RED_HEX
                        "orange" -> return ORANGE_HEX
                        "yellow" -> return YELLOW_HEX
                    }
                }
            }
            return when (card.optString("severity", "")) {
                "severe" -> RED_HEX
                "warning" -> ORANGE_HEX
                "watch" -> YELLOW_HEX
                else -> INFO_HEX
            }
        }
    }

    /** The exact shape `WidgetSnapshot.fromJson` in Dart reads back. */
    fun toJson(): String {
        val json = JSONObject()
        json.put("v", SCHEMA_VERSION)
        json.put("location", locationName)
        json.put("updated_at", updatedAt ?: JSONObject.NULL)
        json.put("lang", lang)
        json.put("units", units)
        json.put(
            "hero",
            hero?.let {
                JSONObject()
                    .put("temp_c", it.tempC ?: JSONObject.NULL)
                    .put("condition", it.condition)
                    .put("feels_like_c", it.feelsLikeC ?: JSONObject.NULL)
                    .put("tmax_c", it.tmaxC ?: JSONObject.NULL)
                    .put("tmin_c", it.tminC ?: JSONObject.NULL)
            } ?: JSONObject.NULL,
        )
        json.put(
            "pinned",
            pinned?.let {
                JSONObject()
                    .put("type", it.type)
                    .put("title", it.title)
                    .put("insight", it.insight)
                    .put("severity", it.severity)
                    .put("color_hex", it.colorHex)
                    .put("estimated", it.estimated)
            } ?: JSONObject.NULL,
        )
        return json.toString()
    }
}

private fun JSONObject.optStringOrNull(key: String): String? {
    if (isNull(key)) return null
    val value = optString(key, "")
    return value.ifEmpty { null }
}

private fun JSONObject.optDoubleOrNull(key: String): Double? {
    if (isNull(key)) return null
    val value = optDouble(key, Double.NaN)
    return if (value.isNaN()) null else value
}
